server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -88,16 +88,40 @@ public final class ShardRouting implements Writeable, ToXContentObject {
         this.allocationId = allocationId;
         this.expectedShardSize = expectedShardSize;
         this.targetRelocatingShard = initializeTargetRelocatingShard();
-        assert (state == ShardRoutingState.UNASSIGNED && unassignedInfo == null) == false : "unassigned shard must be created with meta";
+
         assert relocationFailureInfo != null : "relocation failure info must be always set";
-        assert (state == ShardRoutingState.UNASSIGNED || state == ShardRoutingState.INITIALIZING) == (recoverySource != null)
-            : "recovery source only available on unassigned or initializing shard but was " + state;
-        assert recoverySource == null || recoverySource == PeerRecoverySource.INSTANCE || primary
-            : "replica shards always recover from primary";
-        assert (currentNodeId == null) == (state == ShardRoutingState.UNASSIGNED)
-            : "unassigned shard must not be assigned to a node " + this;
-        assert relocatingNodeId == null || state == ShardRoutingState.RELOCATING || state == ShardRoutingState.INITIALIZING
-            : state + " shard must not have relocating node " + this;
+        switch (state) {
+            case UNASSIGNED -> {
+                assert currentNodeId == null : state + " shard must not be assigned to a node " + this;
+                assert relocatingNodeId == null : state + " shard must not be relocating to a node " + this;
+                assert unassignedInfo != null : state + " shard must be created with unassigned info " + this;
+                assert recoverySource != null : state + "shard must be created with a recovery source" + this;
+                if (primary == false) {
+                    assert recoverySource == PeerRecoverySource.INSTANCE : "replica shards always recover from primary";
+                }
```

Comment:
Does this work?

```suggestion
                } else {
                    assert recoverySource != null && recoverySource != PeerRecoverySource.INSTANCE : "unassigned primaries do not use peer recoveries";
                }
```

Reply:
This is not true when relocating primary to another node.

Reply:
but a relocating primary is not unassigned?

Reply:
Oh, I was checking this for both initializing and unassigned